### DISCUSSION TITLE
Labelling entities in the training data for NER

### DISCUSSION THREAD

**Discussion by shrinidhin** [explosion/spaCy#9966] [Help: Coding & Implementations] [answered]
Hi. So I am training a custom NER model in spacy for identifying custom entities in my data. Consider the following text:
`Example 1 : Nifty Metal, IT, FMCG,Financial Services and PSU Bank indices`
Now I want the entity labels to be recognized as follows:
```
Nifty Metal                    INDEX
Nifty IT                       INDEX
Nifty FMCG                     INDEX
Nifty Financial Services       INDEX
Nifty PSU Bank                 INDEX
```
`Example 2: NIFTY SMALLCAP 100 AND MIDCAP 100`

To be recognized and labelled as:
```
NIFTY SMALLCAP 100  INDEX
NIFTY MIDCAP 100    INDEX
```
I am having trouble understanding how to tag this data using BILOU tagging approach?Should I instead use an entity ruler and add it as a pattern?

**Comment by polm** [comment, marked as answer]
NER is only for contiguous spans of text. So you can't get output like you want using IOB labels or the NER model (or matcher) alone.

What you can do is annotate the whole span of interest (`Nifty ... indices`) and then use post-processing to manually convert that string to the strings you want. The grammatical structure in your example is known as "coordination" and is pretty complicated, but if you focus on a few kinds of examples you want to deal with you may be able to handle it using dependency information.

**Reply by shrinidhin** [reply]
Alright. Thanks!

**Reply by shrinidhin** [reply]
I also have two other scenarios.
```
Scenario 1 : I have certain entities and their abbreviations
Scenario 2 : Same entities with two different names say a Company named 'ABC' which is also referred to as 'XYZ. 
```
In both these scenarios, I want them to be labelled as the same respective entities. For this case, Adding an Entity Ruler component with these scenarios/patterns added before the NER component should make my model work for all these scenarios. Is that correct?

**Reply by polm** [reply]
So an NER model or an EntityRuler label spans in text and give you information like "this span is a company". They don't tell you if two spans are the same or not. With the EntityRuler you can set an "id" that kind of helps you establish this for exact matches, which may help with your scenarios. But ultimately determining if two things are the same or not happens outside the component, when you use the data in some way - like if you're counting mentions you'll want a canonical key, which you could use the ID for.

**Reply by shrinidhin** [reply]
Cool. Thank you.

**Reply by shrinidhin** [reply]
> So an NER model or an EntityRuler label spans in text and give you information like "this span is a company". They don't tell you if two spans are the same or not. With the EntityRuler you can set an "id" that kind of helps you establish this for exact matches, which may help with your scenarios. But ultimately determining if two things are the same or not happens outside the component, when you use the data in some way - like if you're counting mentions you'll want a canonical key, which you could use the ID for.

Hi, Regarding the abbreviations, I have entities to be recognized with their abbreviations. for example: 

```
Bharat Heavy Electricals (BHEL)
Bharat B-ENTITY
Heavy I-ENTITY
Electricals I-ENTITY
(BHEL) L-ENTITY
```
1. Is the way of tagging correct? Should the abbreviation be tagged separately as an entity?
2. I want it to be able to recognize BHEL as an entity, not (BHEL). So should I modify the training data text to train it without the brackets?

**Reply by polm** [reply]
You should tag the abbreviation separately here. I would expect the brackets to be separate tokens, which should not be considered part of the entity and thus not get a label (or be labeled `O`, equivalently). 

Note that there's no one "correct" way to do things - it really depends on what output you want - but in this case I think what I recommended is pretty standard. 

You might find it helpful to look at Universal Dependencies corpora for a reference of how one project did it.

https://github.com/UniversalDependencies/UD_English-EWT

Also, just to check - all your examples so far have basically just been words of interest with a little bit of surrounding text. Is your input complete sentences, or is it like your examples so far?

**Reply by shrinidhin** [reply]
My input is complete sentences. They are texts from news articles to be precise. I save them into .iob files and tag each token with my custom entity label for training. The other non entity ones are usually tagged 'O'.
I used spacy's tokenizer because I wanted to avoid the brackets or punctuation. If the brackets or hyphens are separate tokens, Will they be counted as delimiters which I will have to eliminate while preparing training data?

**Reply by polm** [reply]
If I were generating `.iob` files I would do something like this:

```
import spacy

nlp = spacy.blank("en") # just need a tokenizer

for tok in nlp("this is a text"):
    print(tok)
```

This way each token will be on its own line, and then you can add a label after it. It is important that you do not delete any tokens here.

I'm not sure what you mean by "counted as delimiters" - brackets are tokens, just like other punctuation or normal words, and they still need IOB tags.

**Reply by shrinidhin** [reply]
Thanks!This is very helpful.The problem with having punctuation tagged separately, is that the tokeniser also splits words like It's as It and 's. So wouldn't that result in a poorly trained entity recognizer?

**Reply by polm** [reply]
Sorry, I didn't see your question since it was edited in. 

In IOB training data, every token needs a label, though for some tokens that label is `O` ("other"). For a word that is an entity in parentheses, you might have labels like this:

```
(    O
Paul     B-PERSON
)     O
```

As long as the tokenizer you use to make the training data and the tokenizer used when training the model are the same, it doesn't matter how the tokenizer works.

I don't really understand what `It's` has to do with anything, sorry. If you have something like `Paul's` then that would be two tokens and only `Paul` would be an entity normally, for example.

**Reply by shrinidhin** [reply]
Got it. This discussion has been very helpful. Thank you very much!
